fence/man/fence_apc_snmp.8
.TH fence_apc_snmp 8

.SH NAME
fence_apc_snmp - I/O Fencing agent for APC PDU SNMP devices

.SH SYNOPSIS
.B
fence_apc_snmp
[\fIOPTION\fR]...

.SH DESCRIPTION
fence_apc is an I/O Fencing agent which can be used with the APC rPDU,
MasterSwitch and MasterSwitch+ network power switch.  It logs into a device via
SNMP and reboots a specified outlet. It supports SNMP v1 and v3 with all combinations
of authenticity/privacy settings.

fence_apc_snmp accepts options on the command line as well as from stdin.
Fenced sends parameters through stdin when it execs the agent.  fence_apc_snmp
can be run by itself with command line options.  This is useful for testing.

.SH OPTIONS
.TP
\fB-a\fP \fIIPaddress\fR
IP address or hostname of the SNMP device. Can be used any syntax supported by snmpget.
.TP
\fB-h\fP
Print out a help message describing available options, then exit.
.TP
\fB-c\fP \fIcommunity\fR
The read/write community string to be used in the request. Default private.
.TP
\fB-n\fP \fIname\fR
Name of port to fence or port index. Port can use syntax switch:port.
.TP
\fB-s\fP \fInumber\fR
The switch to operate on.  Defaults to "1" if not specified.
.TP
\fB-p\fP \fIpassword\fR
Password for login for SNMP v3 (authentication protocol pass phrase).
.TP
\fB-P\fP \fIpassword\fR
Password for privacy for SNMP v3 (privacy protocol password).
.TP
\fB-S\fP \fIscript\fR
Script to run to retrieve password for login for SNMP v3 (authentication protocol pass phrase).
.TP
\fB-R\fP \fIscript\fR
Script to run to retrieve privacy for SNMP v3 (privacy protocol password).
.TP
\fB-l\fP \fIlogin\fR
Login name for SNMP v3 (security name).
.TP
\fB-d\fP \fIversion\fR
SNMP version (1,2c,3). Default 1.
.TP
\fB-b\fP \fIauth_protocol\fR
SNMP authentication protocol (MD5|SHA).
.TP
\fB-E\fP \fIsec_level\fR
SNMP security level (noAuthNoPriv|authNoPriv|authPriv).
.TP
\fB-B\fP \fIpriv_protocol\fR
SNMP privacy protocol (DES|AES).
.TP
\fB-u\fP \fIudp_port\fR
UDP/TCP port to use.
.TP
\fB-o\fP \fIaction\fR
The action required.  off (default), on, status, list or monitor. Deprecated
options (enable -> on and disable -> off) can be used too.
.TP
\fB-v\fP
Verbose. Record session to stdout, or debug file if specified (see -D).
.TP
\fB-D\fP
Specifies file, where will be written debug messages from session.
.TP
\fB-V\fP
Print out a version message, then exit.

.SH STDIN PARAMETERS
.TP
\fIagent = < param >\fR
This option is used by fence_node(8) and is ignored by fence_apc_snmp.
.TP
\fIipaddr = < param >\fR
IP address or hostname of the SNMP device. Can be used any syntax supported by snmpget.
.TP
\fIcommunity = < param >\fR
The read/write community string to be used in the request. Default private.
.TP
\fIport = < param >\fR
Name of port to fence or port index. Port can use syntax switch:port.
.TP
\fIswitch = < param >\fR
The switch to operate on.  Defaults to "1" if not specified.
.TP
\fIpasswd = < param >\fR
Password for login for SNMP v3 (authentication protocol pass phrase).
.TP
\fIsnmp_priv_passwd\fR
Password for privacy for SNMP v3 (privacy protocol password).
.TP
\fIpasswd_script = < param >\fR
Script to run to retrieve password for login for SNMP v3 (authentication protocol pass phrase).
.TP
\fIsnmp_priv_passwd_script = < param>\fR
Password for privacy for SNMP v3 (privacy protocol password).
.TP
\fIlogin = < param >\fR
Login name for SNMP v3 (security name).
.TP
\fIsnmp_version = < param >\fR
SNMP version (1,2c,3). Default 1.
.TP
\fIsnmp_auth_prot = < param >\fR
SNMP authentication protocol (MD5|SHA).
.TP
\fIsnmp_sec_level = < param >\fR
SNMP security level (noAuthNoPriv|authNoPriv|authPriv).
.TP
\fIsnmp_priv_prot = < param >\fR
SNMP privacy protocol (DES|AES).
.TP
\fIudpport = < param >\fR
UDP/TCP port to use.
.TP
\fIaction = < param >\fR
The action required.  off (default), on, status, list or monitor. Deprecated
options (enable -> on and disable -> off) can be used too.
.TP
\fIverbose = < param >\fR
Verbose.  Record session to stdout, or debug file if specified (see debug).
.TP
\fIdebug = < param >\fR
Specifies file, where will be written debug messages from session.

.SH SEE ALSO
fence(8), fence_node(8)


fence/agents/apc_snmp/fence_apc_snmp.py
#!/usr/bin/python

# The Following agent has been tested on:
# - APC Switched Rack PDU (MB:v3.7.0 PF:v2.7.0 PN:apc_hw02_aos_270.bin AF1:v2.7.3 AN1:apc_hw02_aos_270.bin
#    AF1:v2.7.3 AN1:apc_hw02_rpdu_273.bin MN:AP7930 HR:B2) - SNMP v1
# - APC Web/SNMP Management Card (MB:v3.8.6 PF:v3.5.8 PN:apc_hw02_aos_358.bin AF1:v3.5.7 AN1:apc_hw02_aos_358.bin
#    AF1:v3.5.7 AN1:apc_hw02_rpdu_357.bin MN:AP7900 HR:B2) - SNMP v1 and v3 (noAuthNoPrivacy,authNoPrivacy, authPrivacy)
# - APC Switched Rack PDU (MB:v3.7.0 PF:v2.7.0 PN:apc_hw02_aos_270.bin AF1:v2.7.3 AN1:apc_hw02_rpdu_273.bin
#    MN:AP7951 HR:B2) - SNMP v1

import sys, re, pexpect
from fencing import *
from fencing_snmp import *

#BEGIN_VERSION_GENERATION
RELEASE_VERSION="APC SNMP fence agent"
REDHAT_COPYRIGHT=""
BUILD_DATE=""
#END_VERSION_GENERATION

### CONSTANTS ###
# oid defining fence device
OID_SYS_OBJECT_ID='.1.3.6.1.2.1.1.2.0'

### GLOBAL VARIABLES ###
# Device - see ApcRPDU, ApcMSP, ApcMS
device=None

# Port ID
port_id=None
# Switch ID
switch_id=None

# Classes describing Device params
class ApcRPDU:
	# Rack PDU
	status_oid=      '.1.3.6.1.4.1.318.1.1.12.3.5.1.1.4.%d'
	control_oid=     '.1.3.6.1.4.1.318.1.1.12.3.3.1.1.4.%d'
	outlet_table_oid='.1.3.6.1.4.1.318.1.1.12.3.5.1.1.2'
	ident_str="APC rPDU"
	state_on=1
	state_off=2
	turn_on=1
	turn_off=2
	has_switches=False

class ApcMSP:
	# Master Switch+
	status_oid=      '.1.3.6.1.4.1.318.1.1.6.7.1.1.5.%d.1.%d'
	control_oid=     '.1.3.6.1.4.1.318.1.1.6.5.1.1.5.%d.1.%d'
	outlet_table_oid='.1.3.6.1.4.1.318.1.1.6.7.1.1.4'
	ident_str="APC Master Switch+"
	state_on=1
	state_off=2
	turn_on=1
	turn_off=3
	has_switches=True

class ApcMS:
	# Master Switch - seems oldest, but supported on every APC PDU
	status_oid=      '.1.3.6.1.4.1.318.1.1.4.4.2.1.3.%d'
	control_oid=     '.1.3.6.1.4.1.318.1.1.4.4.2.1.3.%d'
	outlet_table_oid='.1.3.6.1.4.1.318.1.1.4.4.2.1.4'
	ident_str="APC Master Switch (fallback)"
	state_on=1
	state_off=2
	turn_on=1
	turn_off=2
	has_switches=False

### FUNCTIONS ###
def apc_set_device(conn,options):
	global device

	agents_dir={'.1.3.6.1.4.1.318.1.3.4.5':ApcRPDU,
		    '.1.3.6.1.4.1.318.1.3.4.4':ApcMSP,
		    None:ApcMS}

	# First resolve type of APC
	apc_type=conn.walk(OID_SYS_OBJECT_ID)

	if (not ((len(apc_type)==1) and (agents_dir.has_key(apc_type[0][1])))):
		apc_type=[[None,None]]

	device=agents_dir[apc_type[0][1]]

	conn.log_command("Trying %s"%(device.ident_str))

def apc_resolv_port_id(conn,options):
	global port_id,switch_id,device

	if (device==None):
		apc_set_device(conn,options)

	# Now we resolv port_id/switch_id
	if ((options["-n"].isdigit()) and ((not device.has_switches) or (options["-s"].isdigit()))):
		port_id=int(options["-n"])

		if (device.has_switches):
			switch_id=int(options["-s"])
	else:
		table=conn.walk(device.outlet_table_oid,30)

		for x in table:
			if (x[1].strip('"')==options["-n"]):
				t=x[0].split('.')
				if (device.has_switches):
					port_id=int(t[len(t)-1])
					switch_id=int(t[len(t)-3])
				else:
					port_id=int(t[len(t)-1])

	if (port_id==None):
		fail_usage("Can't find port with name %s!"%(options["-n"]))

def get_power_status(conn,options):
	global port_id,switch_id,device

	if (port_id==None):
		apc_resolv_port_id(conn,options)

	oid=((device.has_switches) and device.status_oid%(switch_id,port_id) or device.status_oid%(port_id))

	(oid,status)=conn.get(oid)
	return (status==str(device.state_on) and "on" or "off")

def set_power_status(conn, options):
	global port_id,switch_id,device

	if (port_id==None):
		apc_resolv_port_id(conn,options)

	oid=((device.has_switches) and device.control_oid%(switch_id,port_id) or device.control_oid%(port_id))

	conn.set(oid,(options["-o"]=="on" and device.turn_on or device.turn_off))


def get_outlets_status(conn, options):
	global device

	result={}

	if (device==None):
		apc_set_device(conn,options)

	res_ports=conn.walk(device.outlet_table_oid,30)

	for x in res_ports:
		t=x[0].split('.')

		port_num=((device.has_switches) and "%s:%s"%(t[len(t)-3],t[len(t)-1]) or "%s"%(t[len(t)-1]))

                port_name=x[1].strip('"')
                port_status=""
                result[port_num]=(port_name,port_status)

        return result

# Define new options
def apc_snmp_define_defaults():
	all_opt["snmp_version"]["default"]="1"
	all_opt["community"]["default"]="private"

# Main agent method
def main():
	device_opt = [ "help", "version", "agent", "quiet", "verbose", "debug",
		       "action", "ipaddr", "login", "passwd", "passwd_script",
		       "test", "port", "separator", "no_login", "no_password",
		       "snmp_version", "community", "snmp_auth_prot", "snmp_sec_level",
		       "snmp_priv_prot", "snmp_priv_passwd", "snmp_priv_passwd_script",
		       "udpport"]

	atexit.register(atexit_handler)

	apc_snmp_define_defaults()

	options=check_input(device_opt,process_input(device_opt))

        ## Support for -n [switch]:[plug] notation that was used before
	if ((options.has_key("-n")) and (-1 != options["-n"].find(":"))):
		(switch, plug) = options["-n"].split(":", 1)
		if ((switch.isdigit()) and (plug.isdigit())):
		        options["-s"] = switch
			options["-n"] = plug

	if (not (options.has_key("-s"))):
		options["-s"]="1"

	# Operate the fencing device
	fence_action(FencingSnmp(options), options, set_power_status, get_power_status, get_outlets_status)

if __name__ == "__main__":
	main()
